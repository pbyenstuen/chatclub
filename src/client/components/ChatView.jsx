import React, { useState } from 'react';
import useSubmit from "./useSubmit";
import InputField from "./InputField";

export const ChatView = ({ user, sender, recipient, chatLog, sendMessage, api }) => {
    const [message, setMessage] = useState("");

    const { handleSubmit: handleStoreMessage, submitting } = useSubmit(
        async () => {
            await api.messages.storeMessage({ message });
        }, () => {
            return;
        }
    );

    const handleSubmit = (e) => {
        e.preventDefault();
        const username = user.displayName;
        const picture = user.image ? user.image : null;
        sendMessage({ recipient, username, picture, message });
        handleStoreMessage(e);
        setMessage("");
    };

    return (
        <div id="chat-container">
            <header>
                <h2>Chat</h2>
            </header>
            <main id="chat-main">
                {chatLog.map(({ id, username, picture, message }) => (
                    <div key={id} className={user.displayName === username ? "message personal" : "message recipient"}>
                        <div className="message-user-info">
                            {picture && <img src={picture} alt={username} />}
                            <h5>{username}</h5>
                        </div>
                        <p>{message}</p>
                    </div>
                ))}
            </main>
            <footer id="chat-footer">
                <form id="chat-form" onSubmit={handleSubmit}>
                    <InputField
                        className={"chat-input"}
                        autoFocus={true}
                        placeholder={"Write here..."}
                        value={message}
                        onValueChange={setMessage} />
                    <button disabled={submitting}>Send</button>
                </form>
            </footer>
        </div>
    );
};

export default ChatView;
 