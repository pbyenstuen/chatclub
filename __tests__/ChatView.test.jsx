import React from "react";
import ReactDOM from "react-dom";
import { act, Simulate } from "react-dom/test-utils";
import { MemoryRouter } from "react-router";
import ChatView from "../src/client/components/ChatView";

let container;

beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
});

afterEach(() => {
    document.body.removeChild(container);
    container = null;
});

const mountComponent = async (component) => {
    await act(async () => {
        await ReactDOM.render(<MemoryRouter>{component}</MemoryRouter>, container);
    });
}

describe("ChatView", () => {
    it("can render existing messages from chat log", async () => {
        const user = {
            displayName: "Mammy Two Shoes"
        }

        const chatLog = [
            {
                id: "1",
                username: "Tom",
                message: "Cool chat",
            },
            {
                id: "2",
                username: "Jerry",
                message: "^",
            },
        ];

        await mountComponent(<ChatView chatLog={chatLog} user={user} />);

        expect(container).toMatchSnapshot();
        expect(container.querySelector(".message-user-info h5").textContent).toEqual("Tom");
        expect(container.querySelector(".message p").textContent).toEqual("Cool chat");
    });

    it("can send new message", async () => {
        const sendMessage = jest.fn();

        const user = {
            displayName: "Mammy Two Shoes"
        }

        await mountComponent(<ChatView chatLog={[]} user={user} sendMessage={sendMessage} />);

        Simulate.change(container.querySelector("input"), {
            target: { value: "..." },
        });

        Simulate.submit(container.querySelector("form"));

        expect(sendMessage).toBeCalledWith({
            message: "...", 
            picture: null, 
            username: "Mammy Two Shoes"
        });
        expect(container.querySelector("input").getAttribute("value")).toEqual("");
    });
});
