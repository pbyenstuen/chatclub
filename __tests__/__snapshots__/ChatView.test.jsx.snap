// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatView can render existing messages from chat log 1`] = `
<div>
  <div
    id="chat-container"
  >
    <header>
      <h2>
        Chat
      </h2>
    </header>
    <main
      id="chat-main"
    >
      <div
        class="message recipient"
      >
        <div
          class="message-user-info"
        >
          <h5>
            Tom
          </h5>
        </div>
        <p>
          Cool chat
        </p>
      </div>
      <div
        class="message recipient"
      >
        <div
          class="message-user-info"
        >
          <h5>
            Jerry
          </h5>
        </div>
        <p>
          ^
        </p>
      </div>
    </main>
    <footer
      id="chat-footer"
    >
      <form
        id="chat-form"
      >
        <div>
          <label>
            
             
            <input
              class="chat-input"
              name=""
              placeholder="Write here..."
              type="text"
              value=""
            />
          </label>
        </div>
        <button>
          Send
        </button>
      </form>
    </footer>
  </div>
</div>
`;
